c/tskit/tables.c

```diff
@@ -5388,6 +5381,638 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+    }
+
+    // Replace the default values if an older parent node is found.
+    if (num_edges > 0) {
+        for (j = 0; j < (int) num_edges; j++) {
+            c = input_edges->child[j];
+            p = &self->oldest_parent[c];
+            if (node_times[*p] < node_times[(int) input_edges->parent[j]]) {
+                self->oldest_parent[c] = input_edges->parent[j];
+            }
+            assert(&self->oldest_parent[c] != NULL);
+        }
+    }
+}
+
+static int
+ibd_finder_init(tsk_ibd_finder_t *self, tsk_id_t *pairs, size_t num_pairs,
+    // tsk_id_t *samples, size_t num_samples,
+    tsk_table_collection_t *tables)
+{
+    int ret = 0;
+    size_t num_nodes_alloc;
+
+    memset(self, 0, sizeof(tsk_ibd_finder_t));
+    // self->samples = samples;
+    // self->num_samples = num_samples;
+    self->pairs = pairs;
+    self->num_pairs = num_pairs;
+    // Should this assertion go here?
+    self->sequence_length = tables->sequence_length;
+    self->num_nodes = tables->nodes.num_rows;
+    self->tables = tables;
+    // self->num_pairs = num_samples * (num_samples - 1) / 2;
+
+    if (pairs == NULL || num_pairs < 1) {
+        ret = TSK_ERR_NO_SAMPLE_PAIRS;
+        goto out;
+    }
+    // Do we need something like this here, or in the Python interface?
+    // assert(sizeof(*pairs)/sizeof(pairs[0]) == 2*sizeof(num_pairs)/sizeof(size_t));
+
+    // Allocate the heaps used for small objects.
+    ret = tsk_blkalloc_init(&self->segment_heap, 8192);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // Mallocing and callocing.
+    num_nodes_alloc = 1 + tables->nodes.num_rows;
+    self->ancestor_map_head = calloc(num_nodes_alloc, sizeof(tsk_segment_t *));
+    self->ancestor_map_tail = calloc(num_nodes_alloc, sizeof(tsk_segment_t *));
+    self->ibd_segments_head = calloc(self->num_pairs, sizeof(tsk_segment_t *));
+    self->ibd_segments_tail = calloc(self->num_pairs, sizeof(tsk_segment_t *));
+    self->sample_id_map = malloc(num_nodes_alloc * sizeof(tsk_id_t));
+    self->samples = malloc(num_nodes_alloc * sizeof(tsk_id_t));
+    self->is_sample = calloc(num_nodes_alloc, sizeof(bool));
+    self->segment_queue_size = 0;
+    self->max_segment_queue_size = 64;
+    self->segment_queue = malloc(self->max_segment_queue_size * sizeof(tsk_segment_t));
+    self->oldest_parent = calloc(num_nodes_alloc, sizeof(tsk_id_t *));
+    if (self->ancestor_map_head == NULL || self->ancestor_map_tail == NULL
+        || self->ibd_segments_head == NULL || self->ibd_segments_tail == NULL
+        || self->sample_id_map == NULL || self->is_sample == NULL
+        || self->segment_queue == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    // Do we need the below? Seems like no?
+    // memset(self->sample_id_map, 0xff, self->tables.nodes.num_rows * sizeof(tsk_id_t));
+    ret = ibd_finder_init_samples(self);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // TODO: feels weird not having a return statement for this?
```

Comment:
It looks fine - it doesn't have any error conditions, so no need for return value.